Report loading status for executable dictionaries correctly

tonickkozlov opened the pull request:
# Changelog category (leave one):
* Bug Fix (user-visible misbehavior in an official stable release)

# Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Report correct status (FAILED) for executable dictionaries that failed to load.

Fixes #37572

tonickkozlov commented:
A real use-case for this change is doing a python2 to python3. If you e.g. start using a base image that only ships with python3, all the dictionaries that rely on python2 will break quietly.

evillique commented on src/Common/waitForPid.cpp:
```diff
@@ -157,7 +157,7 @@ static PollPidResult pollPid(pid_t pid, int timeout_in_ms)
 #endif
 
-bool waitForPid(pid_t pid, size_t timeout_in_seconds)
+bool waitForPid(pid_t pid, size_t timeout_in_seconds, int* out_status)
```
```suggestion
bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_status)
```

evillique commented on src/Common/waitForPid.h:
```diff
@@ -7,6 +7,6 @@ namespace DB
  * Returns `true` if process terminated successfully in specified timeout or `false` otherwise
  */
-bool waitForPid(pid_t pid, size_t timeout_in_seconds);
+bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_pid = nullptr);
```
```suggestion
bool waitForPid(pid_t pid, size_t timeout_in_seconds, int * out_status = nullptr);
```

evillique reviewed (commented):
Please take a look at integration tests `test_executable_dictionary` and the new `test_dictionaries_broken_config` as they fail in the CI.

tonickkozlov commented:
Thanks for your review @evillique , updated my PR.

tonickkozlov committed:
Report loading status for executable dictionaries correctly

robot-ch-test-poll commented:
<!-- automatic status comment for PR #48775 from tonickkozlov/ClickHouse:tonickkozlov/22.3-unexpected-dictionary-load-behaviour -->
This is an automated comment for commit 5b7b424d634ae614d0ec450526454ee9074f88c2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48775/5b7b424d634ae614d0ec450526454ee9074f88c2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

alexey-milovidov commented:
@tonickkozlov LGTM, and the tests passed.

But I didn't understand whether we needed a new configuration parameter. Why don't always check?
And if we will end up introducing a new configuration parameter, it should be covered by docs.

tonickkozlov commented:
@alexey-milovidov Thank you for taking a look. I also think that checking for exit status code should be the default behaviour, no configuration parameter required.
When I tried that previously it broke tests regarding default dictionary column values. My understanding is that the default behaviour is not reporting the executable error but silently filling in the default values. I have just made a change to always check exit statuses so we can verify that's the case.
I'm not sure how to consolidate the two behaviours (reporting the error and filling in the default values). Maybe someone from your team could help? 🙏  Really appreciate it!

alexey-milovidov commented:
I don't think filling the default values was intended. Maybe change it unconditionally, and it will be alright?

tonickkozlov commented:
@alexey-milovidov I've simplified the code and updated the integration test to reflect the desired behaviour.
`00534_functions_bad_arguments` stateless test is failing in msan, but I can't see why. Could it be related? Thank you.

alexey-milovidov commented:
Ok. One question - do we need the "strict_exit_code_check" parameter? If we do, can we rename it to "check_exit_code", because, if I understood correctly, it means "check the exit code".

alexey-milovidov commented:
@tonickkozlov one question remains.

tonickkozlov commented:
@alexey-milovidov I think the parameter can be useful to control behaviour more granularly. For example, UDFs also rely on the ShellCommandSource, and I don't want to break the current default behaviour (falling back to default) in case some users depend on it. So, my change affects only dictionaries. Does it make sense?

alexey-milovidov commented:
Yes, all good.

I'll review the "metadata_loading" test and we will merge.

tonickkozlov commented:
Thank you for quick turn-around @alexey-milovidov 🙏
